README.md
English | [简体中文](./README-CN.md)


<p align="center">
<a href=" https://www.alibabacloud.com"><img src="https://aliyunsdk-pages.alicdn.com/icons/AlibabaCloud.svg"></a>
</p>

<h1 align="center">Alibaba Cloud Core SDK for Ruby(Working in progress)</h1>

<p align="center">
<a href="https://badge.fury.io/rb/aliyunsdkcore"><img src="https://badge.fury.io/rb/aliyunsdkcore.svg" alt="Gem Version"></a>
<a href="https://travis-ci.org/aliyun/openapi-core-ruby-sdk"><img src="https://travis-ci.org/aliyun/openapi-core-ruby-sdk.svg?branch=master" alt="Build Status"></a>
<a href="https://ci.appveyor.com/project/aliyun/openapi-core-ruby-sdk/branch/master"><img src="https://ci.appveyor.com/api/projects/status/uyepkk5bjbynofvu/branch/master?svg=true" alt="Build status"></a>
<a href="https://codecov.io/gh/aliyun/openapi-core-ruby-sdk"><img src="https://codecov.io/gh/aliyun/openapi-core-ruby-sdk/branch/master/graph/badge.svg" alt="codecov"></a>
</p>


Alibaba Cloud Core SDK for Ruby allows you to access Alibaba Cloud services such as Elastic Compute Service (ECS), Server Load Balancer (SLB), and CloudMonitor. You can access Alibaba Cloud services without the need to handle API related tasks, such as signing and constructing your requests.

This document introduces how to install and use Alibaba Cloud Core SDK for Ruby.

## Installation

```sh
$ gem install aliyunsdkcore
```

## Usage

The RPC demo:

```ruby
require 'aliyunsdkcore'

client = RPCClient.new(
  endpoint:          'http://ros.aliyuncs.com',
  api_version:       '2015-09-01',
  access_key_id:     ENV['ACCESS_KEY_ID'],
  access_key_secret: ENV['ACCESS_KEY_SECRET'],
)

params         = { key: (1..11).to_a.map(&:to_s) }
request_option = { method: 'POST', timeout: 15000 }
response       = client.request(
  action: 'DescribeRegions',
  params: params,
  opts: request_option
)

print JSON.parse(response.body)
```


The ROA demo:

```ruby
require 'aliyunsdkcore'

client = ROAClient.new(
  endpoint:          'http://ros.aliyuncs.com',
  api_version:       '2015-09-01',
  access_key_id:     ENV['ACCESS_KEY_ID'],
  access_key_secret: ENV['ACCESS_KEY_SECRET'],
)

response = client.request(
  method: 'GET',
  uri: '/regions',
  options: {
    timeout: 15000
  }
)

print JSON.parse(response.body)
```


## Issues
[Opening an Issue](https://github.com/aliyun/openapi-core-ruby-sdk/issues/new/choose), Issues not conforming to the guidelines may be closed immediately.


## Changelog
Detailed changes for each release are documented in the [release notes](CHANGELOG.md).


## Contribution
Please make sure to read the [Contributing Guide](CONTRIBUTING.md) before making a pull request.


## License
[MIT](LICENSE.md)

Copyright 1999-2019 Alibaba Group Holding Ltd.


CHANGELOG.md
# CHANGELOG

## 1.0.0 - 2019-04
- Waiting for release.


spec/rpc_spec.rb
require "rpc"

RSpec.describe 'RPC' do

#   describe('RPC private methods', function () {
#     const rpc = rewire('../lib/rpc');

  describe 'RPC private methods' do
    it 'firstLetterUpper should ok' do
      expect(firstLetterUpper('helloWorld')).to eq("HelloWorld")
    end

    # it 'formatParams should ok' do
    #   expect(formatParams({'hello': 'world'})).to eql({
    #     Hello: 'world'
    #   })
    # end

    it 'pad2 should ok' do
      expect(pad2(1)).to eq('01')
      expect(pad2(10)).to eq('10')
    end

    it 'timestamp should ok' do
      expect(timestamp).to match(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z/)
    end

#     it('encode should ok', function () {
#       const encode = rpc.__get__('encode');
#       expect(encode('str')).to.be('str');
#       expect(encode('str\'str')).to.be('str%27str');
#       expect(encode('str(str')).to.be('str%28str');
#       expect(encode('str)str')).to.be('str%29str');
#       expect(encode('str*str')).to.be('str%2Astr');
#     });

#     it('replaceRepeatList should ok', function () {
#       const replaceRepeatList = rpc.__get__('replaceRepeatList');
#       function helper(target, key, repeat) {
#         replaceRepeatList(target, key, repeat);
#         return target;
#       }
#       expect(helper({}, 'key', [])).to.be.eql({});
#       expect(helper({}, 'key', ['value'])).to.be.eql({
#         'key.1': 'value'
#       });
#       expect(helper({}, 'key', [{
#         Domain: '1.com'
#       }])).to.be.eql({
#         'key.1.Domain': '1.com'
#       });
#     });

#     it('flatParams should ok', function () {
#       const flatParams = rpc.__get__('flatParams');
#       expect(flatParams({})).to.be.eql({});
#       expect(flatParams({key: ['value']})).to.be.eql({
#         'key.1': 'value'
#       });
#       expect(flatParams({
#         'key': 'value'
#       })).to.be.eql({
#         'key': 'value'
#       });
#       expect(flatParams({
#         key: [
#           {
#             Domain: '1.com'
#           }
#         ]
#       })).to.be.eql({
#         'key.1.Domain': '1.com'
#       });
#     });

#     it('normalize should ok', function () {
#       const normalize = rpc.__get__('normalize');
#       expect(normalize({})).to.be.eql([]);
#       expect(normalize({key: ['value']})).to.be.eql([
#         ['key.1', 'value']
#       ]);
#       expect(normalize({
#         'key': 'value'
#       })).to.be.eql([
#         ['key', 'value']
#       ]);
#       expect(normalize({
#         key: [
#           {
#             Domain: '1.com'
#           }
#         ]
#       })).to.be.eql([
#         ['key.1.Domain', '1.com']
#       ]);
#       expect(normalize({
#         'a': 'value',
#         'c': 'value',
#         'b': 'value'
#       })).to.be.eql([
#         ['a', 'value'],
#         ['b', 'value'],
#         ['c', 'value']
#       ]);
#     });

#     it('canonicalize should ok', function () {
#       const canonicalize = rpc.__get__('canonicalize');
#       expect(canonicalize([])).to.be('');
#       expect(canonicalize([
#         ['key.1', 'value']
#       ])).to.be('key.1=value');
#       expect(canonicalize([
#         ['key', 'value']
#       ])).to.be('key=value');
#       expect(canonicalize([
#         ['key.1.Domain', '1.com']
#       ])).to.be('key.1.Domain=1.com');
#       expect(canonicalize([
#         ['a', 'value'],
#         ['b', 'value'],
#         ['c', 'value']
#       ])).to.be('a=value&b=value&c=value');
#     });
#   });
  end
end

# describe('rpc core', function() {
#   describe('RPCClient', function() {
#     it('should pass into "config"', function() {
#       expect(function () {
#         new RPCClient();
#       }).to.throwException(/must pass "config"/);
#     });

#     it('should pass into "config.endpoint"', function() {
#       expect(function () {
#         new RPCClient({});
#       }).to.throwException(/must pass "config\.endpoint"/);
#     });

#     it('should pass into valid "config.endpoint"', function() {
#       expect(function () {
#         new RPCClient({
#           endpoint: 'ecs.aliyuncs.com/'
#         });
#       }).to.throwException(/"config\.endpoint" must starts with 'https:\/\/' or 'http:\/\/'\./);
#     });

#     it('should pass into "config.apiVersion"', function() {
#       expect(function () {
#         new RPCClient({
#           endpoint: 'http://ecs.aliyuncs.com/'
#         });
#       }).to.throwException(/must pass "config\.apiVersion"/);
#     });

#     it('should pass into "config.accessKeyId"', function() {
#       expect(function () {
#         new RPCClient({
#           endpoint: 'http://ecs.aliyuncs.com/',
#           apiVersion: '1.0'
#         });
#       }).to.throwException(/must pass "config\.accessKeyId"/);
#     });

#     it('should pass into "config.accessKeySecret"', function() {
#       expect(function () {
#         new RPCClient({
#           endpoint: 'http://ecs.aliyuncs.com/',
#           apiVersion: '1.0',
#           accessKeyId: 'accessKeyId'
#         });
#       }).to.throwException(/must pass "config\.accessKeySecret"/);
#     });

#     it('should ok with http endpoint', function() {
#       const client = new RPCClient({
#         endpoint: 'http://ecs.aliyuncs.com',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret'
#       });
#       expect(client.endpoint).to.be('http://ecs.aliyuncs.com');
#       expect(client.keepAliveAgent.protocol).to.be('http:');
#     });

#     it('should ok with https endpoint', function() {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret'
#       });
#       expect(client.endpoint).to.be('https://ecs.aliyuncs.com');
#       expect(client.keepAliveAgent.protocol).to.be('https:');
#     });

#     it('should ok with codes', function() {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#         codes: ['True']
#       });
#       expect(client.codes.has('True')).to.be.ok();
#     });
#   });

#   describe('_buildParams', function() {
#     it('should ok', function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret'
#       });
#       const defaults = client._buildParams();
#       expect(defaults).to.only.have.keys('Format', 'SignatureMethod',
#         'SignatureNonce', 'SignatureVersion', 'Timestamp', 'AccessKeyId',
#         'Version');
#     });

#     it('should ok with securityToken', function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#         securityToken: 'securityToken'
#       });
#       const defaults = client._buildParams();
#       expect(defaults).to.only.have.keys('Format', 'SignatureMethod',
#         'SignatureNonce', 'SignatureVersion', 'Timestamp', 'AccessKeyId',
#         'Version', 'SecurityToken');
#     });
#   });

#   function mock(response, body) {
#     before(function() {
#       muk(httpx, 'request', function(url, opts) {
#         return Promise.resolve(response);
#       });

#       muk(httpx, 'read', function(response, encoding) {
#         return Promise.resolve(body);
#       });
#     });

#     after(function () {
#       muk.restore();
#     });
#   }

#   describe('request', function () {
#     mock({
#       req: {
#         _headers: {}
#       }
#     }, '{}');

#     it('get with raw body should ok', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#       });
#       const result = await client.request('action', {});
#       expect(result).to.be.eql({});
#     });
#   });

#   describe('request with post', function () {
#     mock({
#       req: {
#         _headers: {}
#       },
#       statusCode: 200,
#       headers: {}
#     }, '{}');

#     it('should ok', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#       });
#       const result = await client.request('action');
#       expect(result).to.be.eql({});
#     });

#     it('should ok with formatAction', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#       });
#       const result = await client.request('action', {}, {
#         formatAction: false
#       });
#       expect(result).to.be.eql({});
#     });

#     it('should ok with formatParams', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#       });
#       const result = await client.request('action', {}, {
#         formatParams: false
#       });
#       expect(result).to.be.eql({});
#     });

#     it('should ok with formatParams', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#       });
#       const result = await client.request('action', {}, {
#         agent: new require('https').Agent({
#           keepAlive: true,
#           keepAliveMsecs: 3000
#         })
#       });
#       expect(result).to.be.eql({});
#     });

#     it('get with raw body should ok', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#       });
#       const result = await client.request('action', {}, {
#         method: 'POST'
#       });
#       expect(result).to.be.eql({});
#     });

#     it('get with verbose should ok', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret'
#       }, true);
#       const [json, entry] = await client.request('action', {}, {});
#       expect(json).to.be.eql({});
#       expect(entry.request).to.be.eql({
#         headers: {}
#       });
#       expect(entry.response).to.be.eql({
#         statusCode: 200,
#         headers: {}
#       });
#     });
#   });

#   describe('request with error', function () {
#     mock({
#       req: {
#         _headers: {}
#       }
#     }, JSON.stringify({
#       Code: '400',
#       Message: 'error message'
#     }));

#     it('request with 400 should ok', async function () {
#       const client = new RPCClient({
#         endpoint: 'https://ecs.aliyuncs.com/',
#         apiVersion: '1.0',
#         accessKeyId: 'accessKeyId',
#         accessKeySecret: 'accessKeySecret',
#       });
#       try {
#         await client.request('action', {});
#       } catch (ex) {
#         expect(ex.message.startsWith('error message, URL: ')).to.be.ok();
#         return;
#       }
#       // should never be executed
#       expect(false).to.be.ok();
#     });
#   });

#   describe('RPC private methods', function () {
#     const rpc = rewire('../lib/rpc');

#     it('firstLetterUpper should ok', function () {
#       const firstLetterUpper = rpc.__get__('firstLetterUpper');
#       expect(firstLetterUpper('hello')).to.be('Hello');
#     });

#     it('formatParams should ok', function () {
#       const formatParams = rpc.__get__('formatParams');
#       expect(formatParams({'hello': 'world'})).to.be.eql({
#         Hello: 'world'
#       });
#     });

#     it('timestamp should ok', function () {
#       const timestamp = rpc.__get__('timestamp');
#       expect(timestamp()).to.be.match(/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z/);
#     });

#     it('encode should ok', function () {
#       const encode = rpc.__get__('encode');
#       expect(encode('str')).to.be('str');
#       expect(encode('str\'str')).to.be('str%27str');
#       expect(encode('str(str')).to.be('str%28str');
#       expect(encode('str)str')).to.be('str%29str');
#       expect(encode('str*str')).to.be('str%2Astr');
#     });

#     it('replaceRepeatList should ok', function () {
#       const replaceRepeatList = rpc.__get__('replaceRepeatList');
#       function helper(target, key, repeat) {
#         replaceRepeatList(target, key, repeat);
#         return target;
#       }
#       expect(helper({}, 'key', [])).to.be.eql({});
#       expect(helper({}, 'key', ['value'])).to.be.eql({
#         'key.1': 'value'
#       });
#       expect(helper({}, 'key', [{
#         Domain: '1.com'
#       }])).to.be.eql({
#         'key.1.Domain': '1.com'
#       });
#     });

#     it('flatParams should ok', function () {
#       const flatParams = rpc.__get__('flatParams');
#       expect(flatParams({})).to.be.eql({});
#       expect(flatParams({key: ['value']})).to.be.eql({
#         'key.1': 'value'
#       });
#       expect(flatParams({
#         'key': 'value'
#       })).to.be.eql({
#         'key': 'value'
#       });
#       expect(flatParams({
#         key: [
#           {
#             Domain: '1.com'
#           }
#         ]
#       })).to.be.eql({
#         'key.1.Domain': '1.com'
#       });
#     });

#     it('normalize should ok', function () {
#       const normalize = rpc.__get__('normalize');
#       expect(normalize({})).to.be.eql([]);
#       expect(normalize({key: ['value']})).to.be.eql([
#         ['key.1', 'value']
#       ]);
#       expect(normalize({
#         'key': 'value'
#       })).to.be.eql([
#         ['key', 'value']
#       ]);
#       expect(normalize({
#         key: [
#           {
#             Domain: '1.com'
#           }
#         ]
#       })).to.be.eql([
#         ['key.1.Domain', '1.com']
#       ]);
#       expect(normalize({
#         'a': 'value',
#         'c': 'value',
#         'b': 'value'
#       })).to.be.eql([
#         ['a', 'value'],
#         ['b', 'value'],
#         ['c', 'value']
#       ]);
#     });

#     it('canonicalize should ok', function () {
#       const canonicalize = rpc.__get__('canonicalize');
#       expect(canonicalize([])).to.be('');
#       expect(canonicalize([
#         ['key.1', 'value']
#       ])).to.be('key.1=value');
#       expect(canonicalize([
#         ['key', 'value']
#       ])).to.be('key=value');
#       expect(canonicalize([
#         ['key.1.Domain', '1.com']
#       ])).to.be('key.1.Domain=1.com');
#       expect(canonicalize([
#         ['a', 'value'],
#         ['b', 'value'],
#         ['c', 'value']
#       ])).to.be('a=value&b=value&c=value');
#     });
#   });
# });

spec/hello_spec.rb
# in spec/hello_spec.rb
require "hello"

RSpec.describe do
  describe 'hello' do
    it 'returns Hello, Ruby.' do
      expect(hello).to eq("Hello, Ruby.")
    end
  end
end

spec/spec_helper.rb
require 'simplecov'
SimpleCov.start do
  add_filter "/spec/"
end

if ENV['CI'] == 'true'
  require 'codecov'
  SimpleCov.formatter = SimpleCov::Formatter::Codecov
end

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end

lib/roa_client.rb
require 'faraday'
require 'securerandom'
require 'active_support/all'

class ROAClient

  attr_accessor :endpoint, :api_version, :access_key_id, :access_key_secret, :security_token, :hostname, :opts

  def initialize(config)

    validate config

    self.endpoint          = config[:endpoint]
    self.api_version       = config[:api_version]
    self.access_key_id     = config[:access_key_id]
    self.access_key_secret = config[:access_key_secret]
    self.security_token    = config[:security_token]

  end

  def request(method:, uri:, params: {}, body: {}, headers: {}, options: {})

    mix_headers = default_headers.merge(headers)

    response = connection.send(method.downcase) do |request|
      request.url uri, params
      if body.present?
        request_body                  = body.to_json
        request.body                  = request_body
        mix_headers['content-md5']    = Digest::MD5.base64digest request_body
        mix_headers['content-length'] = request_body.length.to_s
      end
      string_to_sign = string_to_sign(method, uri, mix_headers, params)
      mix_headers.merge!(authorization: authorization(string_to_sign))
      mix_headers.each { |key, value| request.headers[key] = value }
    end

    return response if options.has_key? :raw_body

    response_content_type = response.headers['Content-Type'] || ''
    if response_content_type.start_with?('application/json')
      if response.status >= 400
        result = JSON.parse(response.body)
        raise StandardError, "code: #{response.status}, #{result['Message']} requestid: #{result['RequestId']}"
      end
    end

    if response_content_type.start_with?('text/xml')
      result = Hash.from_xml(response.body)
      raise ACSError, result['Error'] if result['Error']
    end

    response
  end

  def connection(adapter = Faraday.default_adapter)
    Faraday.new(:url => self.endpoint) { |faraday| faraday.adapter adapter }
  end

  def get(uri: '', headers: {}, params: {}, options: {})
    request(method: :get, uri: uri, params: params, body: {}, headers: headers, options: options)
  end

  def post(uri: '', headers: {}, params: {}, body: {}, options: {})
    request(method: :get, uri: uri, params: params, body: body, headers: headers, options: options)
  end

  def put(uri: '', headers: {}, params: {}, body: {}, options: {})
    request(method: :get, uri: uri, params: params, body: body, headers: headers, options: options)
  end

  def delete(uri: '', headers: {}, params: {}, options: {})
    request(method: :get, uri: uri, params: params, body: {}, headers: headers, options: options)
  end

  def default_headers
    default_headers = {
      'accept':                  'application/json',
      'date':                    Time.now.httpdate,
      'host':                    URI(self.endpoint).host,
      'x-acs-signature-nonce':   SecureRandom.hex(16),
      'x-acs-signature-method':  'HMAC-SHA1',
      'x-acs-signature-version': '1.0',
      'x-acs-version':           self.api_version,
      'x-sdk-client':            "RUBY(#{RUBY_VERSION})" # FIXME: 如何获取Gem的名称和版本号
    }
    if self.security_token
      default_headers.merge!('x-acs-accesskey-id': self.access_key_id, 'x-acs-security-token': self.security_token)
    end
    default_headers
  end

  private

  def string_to_sign(method, uri, headers, query = {})
    headers.stringify_keys!
    header_string = [
      method,
      headers['accept'],
      headers['content-md5'] || '',
      headers['content-type'] || '',
      headers['date'],
    ].join("\n")
    "#{header_string}\n#{canonicalized_headers(headers)}#{canonicalized_resource(uri, query)}"
  end

  def canonicalized_headers(headers)
    headers.keys.select { |key| key.to_s.start_with? 'x-acs-' }
      .sort.map { |key| "#{key}:#{headers[key].strip}\n" }.join
  end

  def canonicalized_resource(uri, query_hash = {})
    query_string = query_hash.map { |key, value| "#{key}=#{value}" }.join('&')
    query_string.empty? ? uri : "#{uri}?#{query_string}"
  end

  def authorization(string_to_sign)
    "acs #{self.access_key_id}:#{signature(string_to_sign)}"
  end

  def signature(string_to_sign)
    Base64.encode64(OpenSSL::HMAC.digest('sha1', self.access_key_secret, string_to_sign)).strip
  end

  def validate(config)
    raise ArgumentError, 'must pass "config"' unless config
    raise ArgumentError, 'must pass "config[:endpoint]"' unless config[:endpoint]
    unless config[:endpoint].start_with?('http://') || config[:endpoint].start_with?('https://')
      raise ArgumentError, '"config.endpoint" must starts with \'https://\' or \'http://\'.'
    end
    raise ArgumentError, 'must pass "config[:api_version]"' unless config[:api_version]
    raise ArgumentError, 'must pass "config[:access_key_id]"' unless config[:access_key_id]
    raise ArgumentError, 'must pass "config[:access_key_secret]"' unless config[:access_key_secret]
  end

  class ACSError < StandardError

    attr_accessor :code

    def initialize(error)
      self.code  = error['Code']
      message    = error['Message']
      host_id    = error['HostId']
      request_id = error['RequestId']
      super("#{message} host_id: #{host_id}, request_id: #{request_id}")
    end

  end

end


lib/rpc.rb
# const assert = require('assert');

# const httpx = require('httpx');
# const kitx = require('kitx');
# const JSON = require('json-bigint');
require 'date'

def firstLetterUpper(str)
  return str[0].upcase + str[1..-1]
end

# def formatParams(params)
#   keys = Object.keys(params);
#   newParams = {};
#   for (var i = 0; i < keys.length; i++) {
#     var key = keys[i];
#     newParams[firstLetterUpper(key)] = params[key];
#   }
#   return newParams;
# end

def pad2(n)
  if n < 10 then
    return '0' + String(n)
  else
    return String(n)
  end
end

def timestamp
  date = Time.now
  _YYYY = date.year
  _MM = pad2(date.month)
  _DD = pad2(date.day)
  _HH = pad2(date.hour)
  mm = pad2(date.min)
  ss = pad2(date.sec)
  # 删除掉毫秒部分
  return "#{_YYYY}-#{_MM}-#{_DD}T#{_HH}:#{mm}:#{ss}Z";
end

# def encode(str)
#   result = encodeURIComponent(str);

#   return result.replace(/\!/g, '%21')
#    .replace(/\'/g, '%27')
#    .replace(/\(/g, '%28')
#    .replace(/\)/g, '%29')
#    .replace(/\*/g, '%2A');
# end

# def replaceRepeatList(target, key, repeat)
#   for item in repeat
#     if (item && typeof item === 'object') then
#       const keys = Object.keys(item);
#       for currentKey in keys
#         target[`${key}.${i + 1}.${currentKey}`] = item[currentKey];
#       }
#     else
#       target[`${key}.${i + 1}`] = item;
#     end
#   end
# end

# def flatParams(params)
#   var target = {};
#   var keys = Object.keys(params);
#   for (let i = 0; i < keys.length; i++) {
#     var key = keys[i];
#     var value = params[key];
#     if (Array.isArray(value)) then
#       replaceRepeatList(target, key, value);
#     else
#       target[key] = value;
#     end
#   }
#   return target;
# end

# def normalize(params)
#   var list = [];
#   var flated = flatParams(params);
#   var keys = Object.keys(flated).sort();
#   for (let i = 0; i < keys.length; i++) {
#     var key = keys[i];
#     var value = flated[key];
#     list.push([encode(key), encode(value)]); #push []
#   }
#   return list;
# end

# def canonicalize(normalized)
#   var fields = [];
#   for (var i = 0; i < normalized.length; i++) {
#     var [key, value] = normalized[i];
#     fields.push(key + '=' + value);
#   }
#   return fields.join('&');
# end

# class RPCClient
#   def initialize(config, verbose)
#     assert(config, 'must pass "config"');
#     assert(config.endpoint, 'must pass "config.endpoint"');
#     if (!config.endpoint.startsWith('https:#') &&
#       !config.endpoint.startsWith('http:#')) then
#       throw new Error(`"config.endpoint" must starts with 'https:#' or 'http:#'.`);
#     end
#     assert(config.apiVersion, 'must pass "config.apiVersion"');
#     assert(config.accessKeyId, 'must pass "config.accessKeyId"');
#     accessKeySecret = config.secretAccessKey || config.accessKeySecret;
#     assert(accessKeySecret, 'must pass "config.accessKeySecret"');

#     if (config.endpoint.endsWith('/')) then
#       config.endpoint = config.endpoint.slice(0, -1);
#     end

#     @endpoint = config.endpoint;
#     @apiVersion = config.apiVersion;
#     @accessKeyId = config.accessKeyId;
#     @accessKeySecret = accessKeySecret;
#     @securityToken = config.securityToken;
#     @verbose = verbose === true;
#     # 非 codes 里的值，将抛出异常
#     @codes = new Set([200, '200', 'OK', 'Success']);
#     if (config.codes) then
#       # 合并 codes
#       for (var elem of config.codes) {
#         @codes.add(elem);
#       }
#     end

#     @opts = config.opts || {};

#     httpModule = @endpoint.startsWith('https:#')
#       ? require('https') : require('http');
#     @keepAliveAgent = new httpModule.Agent({
#       keepAlive: true,
#       keepAliveMsecs: 3000
#     });
#   end

#   def request(action, params = {}, opts = {})
#     # 1. compose params and opts
#     opts = Object.assign({}, @opts, opts);

#     # format action until formatAction is false
#     if (opts.formatAction !== false) then
#       action = firstLetterUpper(action);
#     end

#     # format params until formatParams is false
#     if (opts.formatParams !== false) then
#       params = formatParams(params);
#     end
#     var defaults = @_buildParams();
#     params = Object.assign({Action: action}, defaults, params);

#     # 2. caculate signature
#     var method = (opts.method || 'GET').toUpperCase();
#     var normalized = normalize(params);
#     var canonicalized = canonicalize(normalized);
#     # 2.1 get string to sign
#     var stringToSign = `${method}&${encode('/')}&${encode(canonicalized)}`;
#     # 2.2 get signature
#     const key = @accessKeySecret + '&';
#     var signature = kitx.sha1(stringToSign, key, 'base64');
#     # add signature
#     normalized.push(['Signature', encode(signature)]);
#     # 3. generate final url
#     const url = opts.method === 'POST' ? `${@endpoint}/` : `${@endpoint}/?${canonicalize(normalized)}`;
#     # 4. send request
#     var entry = {
#       url: url,
#       request: null,
#       response: null
#     };

#     if (opts && !opts.agent) then
#       opts.agent = @keepAliveAgent;
#     end

#     if (opts.method === 'POST') then
#       opts.headers = {};
#       opts.headers['content-type'] = 'application/x-www-form-urlencoded';
#       opts.data = canonicalize(normalized);
#     end

#     return httpx.request(url, opts).then((response) => {
#       entry.request = {
#         headers: response.req._headers
#       };
#       entry.response = {
#         statusCode: response.statusCode,
#         headers: response.headers
#       };

#       return httpx.read(response);
#     }).then((buffer) => {
#       var json = JSON.parse(buffer);
#       if (json.Code && !@codes.has(json.Code)) {
#         var err = new Error(`${json.Message}, URL: ${url}`);
#         err.name = json.Code + 'Error';
#         err.data = json;
#         err.code = json.Code;
#         err.url = url;
#         err.entry = entry;
#         return Promise.reject(err);
#       }

#       if (@verbose) {
#         return [json, entry];
#       }

#       return json;
#     });
#   end

#   def _buildParams
#     var defaultParams = {
#       Format: 'JSON',
#       SignatureMethod: 'HMAC-SHA1',
#       SignatureNonce: kitx.makeNonce,
#       SignatureVersion: '1.0',
#       Timestamp: timestamp,
#       AccessKeyId: @accessKeyId,
#       Version: @apiVersion,
#     };
#     if (@securityToken) then
#       defaultParams.SecurityToken = @securityToken;
#     end
#     return defaultParams;
#   end
# end
